### PR TITLE
Upgraded Flink on Dataproc support from Interacitve Beam

### PR DESCRIPTION
**Pull request by KevinGG**
* Added initialization actions to configure docker credsHelper with gcloud.
* Improved taskmanager configurations for higher parallelism.
* Fixed a cluster metadata bug that always creates a new cluster instead of reusing the lastly used one if region is not set.
* Added parallelism and max_parallelism Flink pipeline options.
* Changed how Flink-YARN session is started to take effect of configuration changes as well as where to parse Flink Master.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by KevinGG** [comment]
R: @functicons
PTAL, thx!

**Comment by KevinGG** [comment]
R: @rohdesamuel 
PTAL, thx!

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 1]
```diff
@@ -1421,6 +1421,20 @@ def _add_argparse_args(cls, parser):
         ' directly, rather than starting up a job server.'
         ' Only applies when flink_master is set to a'
         ' cluster address.  Requires Python 3.6+.')
+    parser.add_argument(
```
Comment (KevinGG) [inline sdks/python/apache_beam/options/pipeline_options.py:1424]: These mirror the existing [FlinkRunnerOptions.java](https://github.com/apache/beam/blob/master/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java#L66).

**Comment by KevinGG** [comment]
Ran a RunInference example for 50,000 images with 40 n1-highmem-8 workers at 100 parallelism, finished in less than 5 minutes:
parallelism.

**Comment by KevinGG** [comment]
Run PythonLint PreCommit

File: sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py [hunk 2]
```diff
@@ -64,6 +92,30 @@ def __init__(self, cluster_metadata: ClusterMetadata) -> None:
         })
     self._fs = gcsfilesystem.GCSFileSystem(PipelineOptions())
     self._staging_directory = None
+    cache_dir = ie.current_env().options.cache_root
+    try:
+      assert cache_dir.startswith('gs://')
```
Comment (functicons) [inline sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py:97]: assertion is usually used to catch code bug, in this case, it is invalid user input, not a bug. I think it is better to use a if check then throw an exception.

File: sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py [hunk 3]
```diff
@@ -43,7 +45,33 @@ class UnimportedDataproc:
...
+# Configures the os user yarn to use gcloud as the docker credHelper.
+# Also sets some taskmanager configurations for better parallelism.
+# Finally starts the yarn application: flink cluster in session mode.
+INIT_ACTION = """#!/bin/bash
+sudo -u yarn gcloud auth configure-docker --quiet
+
+readonly FLINK_INSTALL_DIR='/usr/lib/flink'
+readonly MASTER_HOSTNAME="$(/usr/share/google/get_metadata_value attributes/dataproc-master)"
+
+cat <<EOF >>${FLINK_INSTALL_DIR}/conf/flink-conf.yaml
```
Comment (functicons) [inline sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py:62]: Do you need to configure jobmanager and taskmanager memory size? The default value set by Dataproc is problematic in some cases AFAIK.

Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py:62]: I don't need to set those values and prefer to let Dataproc select the appropriate values.

The values I set here are basically:
- taskmanager.memory.network.* (somehow Dataproc fixed all values to 64mb and didn't honor the yarn defaults).
- taskmanager.network.numberOfBuffers: this should honor >= `#slots-per-TM^2 * #TMs * 4`. I temporarily increased the default value to support hundreds of parallelism when there are hundreds of CPUs in the cluster. If needed, we shall expose it for customization by end users in the future.

File: sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py (hunk 2 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/dataproc/dataproc_cluster_manager.py:97]: You are absolutely right, assert should not be used to validate user input. I'll change it to raise a ValueError.

**Review by functicons — APPROVED** [review]
LGTM

**Pull request merged**